Loki Query Editor: Autocompletion and suggestions improvements (unwrap, parser, extracted labels)

matyax opened the pull request:
**What is this feature?**

This PR adds several improvements to our autocompletion support for the Monaco query editor.

- Improved `unwrap` identification and suggestion.
- Improved suggestions for `unwrap`: labels or functions.
- Testing and coverage improvements.
- We now send the full log query to get a data sample and extract labels, instead of just the query labels.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/57683

**Special notes for your reviewer**:

Compare previous with current suggestion behavior.

https://user-images.githubusercontent.com/1069378/203310123-d227255e-f1ef-4b82-a871-60bae3248e8f.mov

Demo of sending full logs query for extracted labels:

https://user-images.githubusercontent.com/1069378/203346970-d7106c93-070e-436f-95b9-b1f1bc291dc8.mov

matyax committed:
Chore: refactor test to improve internal categorization of scenarios

matyax committed:
feat(loki-monaco-unwrap): add unwrap situation support

matyax committed:
Chore: remove redundant path from aggregation situation

matyax committed:
feat(loki-monaco-unwrap): add unwrap suggestions

matyax committed:
feat(loki-monaco-autocomplete): rename IN_DURATION and add missing tests

matyax committed:
feat(loki-monaco-autocomplete): detect parser and line filter

matyax committed:
feat(loki-monaco-autocomplete): optionally suggest line filters and parsers

matyax committed:
feat(loki-monaco-autocomplete): improve suggestions and update completions tests

matyax committed:
feat(loki-monaco-autocomplete): allow for multiple line filters suggestions

matyax committed:
Chore: update situations test

matyax committed:
Chore: add test case for AFTER_UNWRAP completion

matyax committed:
feat(loki-monaco-autocomplete): use logs query instead of labels for data sample

matyax committed:
Chore: improve getParser function name and add tests

matyax committed:
Chore: update test mock data

ivanahuckova reviewed (commented):
Really nice job!! 🎉 I tested this and I have couple of questions/thoughts:

> Only suggest parser when it's not present in the log query.

I am wondering if this make sense.  In most cases, you use 1 parser, but there could be cases where you want to use multiple parsers. (I think there could be, but would be nice to check with Loki team). Based on their answer, we can decide  if we want to do show parsers or not. 

> Improved unwrap identification and suggestion.

I am thinking we could move `label_format` and `line_format` above `unwrap`, because now if we have a lot of labels, these options are quite low in the list and hard to find. 

I am also thinking about the fact, that `unwrap` can be used only in metrics query. Therefore I am thinking if we should suggest it if we are in log query, as the resulting query is going to be invalid. Or if we should suggest it when we are in metrics query.

And last thing (I did a video explaining the issue), but when I write pipe `|` I get only `unwrap` suggestion and no labels/functions. 

https://user-images.githubusercontent.com/30407135/204562670-a1af035f-8426-4a76-8821-05aed395d9ef.mov

>Improved suggestions for unwrap: labels or functions.

Here I am wondering, if we should only suggest labels that are number/duration/bytes. Because for all other labels, the resulting query is invalid. And if you use parser, you can have tens of labels, but only fraction would be usable. 

And 2 more non-related-to-this PR issue I noticed - when I write queries:

- `{source="data"} | ^` and I choose parser,  it adds extra space between pipe and parser, so the resulting query is `{source="data"} |  logfmt` instead of  `{source="data"} | logfmt`. 

- `rate({source="data"} |^ [$__interval])` and I choose parser, it adds extra pipe, so the resulting query is `rate({source="data"} | | logfmt[$__interval])` instead of `rate({source="data"} |  logfmt[$__interval])`

matyax committed:
Update public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.test.ts

Co-authored-by: Ivana Huckova <30407135+ivanahuckova@users.noreply.github.com>

matyax commented:
Thanks for the feedback!

> here could be cases where you want to use multiple parsers.

Asked the Loki team.

> move label_format and line_format above unwrap

Good idea, will do that.

> Therefore I am thinking if we should suggest it if we are in log query, as the resulting query is going to be invalid.

That's also a good idea, but I think we can maybe let the query validation handle that?

![imagen](https://user-images.githubusercontent.com/1069378/204862827-2f6b2ed9-d4c3-4290-94f1-ae038c2a3369.png)

>  I write pipe | I get only unwrap suggestion and no labels/functions.

I saw that. The problem is that when you write the pipe the situation changes to `AFTER_SELECTOR`. I'll see if we can improve that, but currently the situation identification struggles with some queries.

matyax commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -256,7 +310,9 @@ export async function getCompletions(
     case 'AFTER_SELECTOR':
-      return getAfterSelectorCompletions(situation.labels, situation.afterPipe, dataProvider);
+      return getAfterSelectorCompletions(situation.logQuery, situation.afterPipe, situation.parser, dataProvider);
+    case 'AFTER_UNWRAP':
```
You get the, but the limitation is that you have to be in the process of writing the query:
`quantile_over_time(0.99, {cluster="ops-tools1",container="ingress-nginx"} | json | __error__ = "" | unwrap `
![imagen](https://user-images.githubusercontent.com/1069378/204864311-4fd3e229-3b43-4966-9fb0-fd72fdd5697d.png)

If you add the closing brace, the situation changes and it no longer identifies it.

`quantile_over_time(0.99, {cluster="ops-tools1",container="ingress-nginx"} | json | __error__ = "" | unwrap )`

![imagen](https://user-images.githubusercontent.com/1069378/204864584-840467f2-5ff5-4060-b8b6-d2b28edb9d77.png)

I'll see if I can improve the situation detection there.

matyax committed:
feat(loki-monaco-autocomplete): improve after unwrap detection

matyax commented:
@ivanahuckova could you please take this for a spin again? Thank you 🙇

ivanahuckova commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:
```diff
@@ -116,7 +119,12 @@ export type Situation =
       type: 'AFTER_SELECTOR';
       afterPipe: boolean;
-      labels: Label[];
+      parser?: string;
```
Do we still need this parser here? I thought that we had it here to not offer 2 parsers, but we learned that there might be many parsers. So do we still need it?

matyax replied:
No, thanks for noticing.

ivanahuckova reviewed (commented):
The line format and label format are still at the bottom 
<img width="666" alt="image" src="https://user-images.githubusercontent.com/30407135/207304548-52d68081-0711-4956-b63c-27a50dc7cbf8.png">


Moreover is there a plan to offer only number/duration/bytes unwrap terms in this PR of in next?

matyax committed:
feat(loki-monaco-autocomplete): remove leftover parser detection

matyax commented:
@ivanahuckova Let's use a follow-up PR to address:
- Move up line_format and label_format 
- Resolve the extra added space.

> Moreover is there a plan to offer only number/duration/bytes unwrap terms in this PR of in next?

It's hard to tell what query are you using there, but we do suggest them in metrics query such as: `sum(sum_over_time({place="luna"} | unwrap ^ [5m])) by (place)

![imagen](https://user-images.githubusercontent.com/1069378/207308178-af0de784-bb5a-419e-8b30-514e2956475f.png)

> and I choose parser, it adds extra pipe, so the resulting query is 

I haven't been able to consistently reproduce this with input. So far it seems to randomly pop up.

matyax commented:
@ivanahuckova 

> I meant, that currently we offer all labels. But unwrap can be only with label that's value is number/duration/bytes type

From https://grafana.com/docs/loki/latest/logql/metric_queries/ I read that 

> The unwrap expression is noted | unwrap label_identifier

And

> Optionally the label identifier can be wrapped by a conversion function | unwrap <function>(label_identifier)

Can you confirm if what you mean is that we're missing autocompletion for `<function>(label_identifier)`?

matyax committed:
Chore: completely remove parser suggestion exclusion implementation

It was correct to have more than one parser, so we don't need to exclude parsers if one is present.

ivanahuckova commented:
No, I meant this part 

```
The unwrap expression is noted | unwrap label_identifier where the label identifier is the label name to use for extracting sample values.

Since label values are string, by default a conversion into a float (64bits) will be attempted, in case of failure the __error__ label is added to the sample. 
``` 

So if we offer `unwrap level` and level is string that can't be converted into float number, user will get error. So we don't want to offer labels that have non-number values, as result of this query is error.

matyax commented:
Oh, I remember that. I'm not really sure if we can resolve that in the autocomplete layer (specially without adding delay as the user types). I'll explore it with the follow up.
